File: backend/src/services/auth.service.ts
import bcrypt from "bcrypt";
import crypto from "crypto";
import prisma from "../utils/prisma";
import { Role } from "../utils/constants";

export class AuthService {
  static async register(username: string, passwordRaw: string, name: string, role?: string) {
    const existingUser = await prisma.user.findUnique({ where: { username } });
    if (existingUser) {
      throw new Error("User already exists");
    }

    const salt = await bcrypt.genSalt(12);
    const passwordHash = await bcrypt.hash(passwordRaw, salt);

    const user = await prisma.user.create({
      data: {
        username,
        passwordHash,
        salt,
        name,
        role: role || Role.DEVELOPER,
      },
    });

    console.log('[AuthService.register] User created:', user.userId);
    return this.createSession(user.userId, user.username, user.role);
  }

  static async login(username: string, passwordRaw: string) {
    const user = await prisma.user.findUnique({ where: { username } });
    if (!user || !user.isActive) {
      throw new Error("Invalid credentials");
    }

    const isPasswordValid = await bcrypt.compare(passwordRaw, user.passwordHash);
    if (!isPasswordValid) {
      throw new Error("Invalid credentials");
    }

    // Delete all expired sessions for this user to keep the table clean
    await prisma.session.deleteMany({
      where: {
        userId: user.userId,
        expiresAt: { lt: new Date() },
      },
    });

    return this.createSession(user.userId, user.username, user.role);
  }

  private static async createSession(userId: string, username: string, role: string) {
    const token = crypto.randomBytes(32).toString('hex');
    const tokenHash = crypto.createHash('sha256').update(token).digest('hex');
    
    // Expires in 7 days
    const expiresAt = new Date();
    expiresAt.setDate(expiresAt.getDate() + 7);

    await prisma.session.create({
      data: {
        userId,
        tokenHash,
        expiresAt
      }
    });

    return { user: { userId, username, role }, token };
  }

  static async getUsers() {
    return prisma.user.findMany({
      select: { userId: true, username: true, name: true, role: true, isActive: true }
    });
  }
}

File: backend/src/main/java/com/patchflow/service/AuthService.java
package com.patchflow.service;

import com.patchflow.entity.Session;
import com.patchflow.entity.User;
import com.patchflow.filter.AuthTokenFilter;
import com.patchflow.repository.SessionRepository;
import com.patchflow.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.security.SecureRandom;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final SessionRepository sessionRepository;
    private final BCryptPasswordEncoder passwordEncoder;

    @Transactional
    public Map<String, Object> login(String username, String password) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials"));

        if (!user.isActive()) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials");
        }

        if (!passwordEncoder.matches(password, user.getPasswordHash())) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials");
        }

        // Clean expired sessions
        sessionRepository.deleteExpiredByUserId(user.getUserId(), Instant.now());

        return createSession(user);
    }

    @Transactional
    public Map<String, Object> register(String username, String password, String name, String role) {
        if (userRepository.existsByUsername(username)) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "User already exists");
        }

        String hash = passwordEncoder.encode(password);

        User user = User.builder()
                .username(username)
                .passwordHash(hash)
                .name(name)
                .role(role != null ? role : "DEVELOPER")
                .isActive(true)
                .build();
        user = userRepository.save(user);
        return createSession(user);
    }

    private Map<String, Object> createSession(User user) {
        // Generate 32-byte random token
        byte[] tokenBytes = new byte[32];
        new SecureRandom().nextBytes(tokenBytes);
        String token = Base64.getUrlEncoder().withoutPadding().encodeToString(tokenBytes);
        String tokenHash = AuthTokenFilter.sha256(token);

        Instant expiresAt = Instant.now().plus(7, ChronoUnit.DAYS);

        Session session = Session.builder()
                .userId(user.getUserId())
                .tokenHash(tokenHash)
                .expiresAt(expiresAt)
                .build();
        sessionRepository.save(session);

        Map<String, Object> userMap = new HashMap<>();
        userMap.put("userId", user.getUserId());
        userMap.put("username", user.getUsername());
        userMap.put("name", user.getName());
        userMap.put("role", user.getRole());
        userMap.put("designation", user.getDesignation());
        userMap.put("email", user.getEmail());
        userMap.put("phone", user.getPhone());

        Map<String, Object> result = new HashMap<>();
        result.put("user", userMap);
        result.put("token", token);
        return result;
    }

    @Transactional
    public void logout(String token) {
        String tokenHash = AuthTokenFilter.sha256(token);
        sessionRepository.deleteByTokenHash(tokenHash);
    }
}
